usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -3642,17 +3642,13 @@ zfs_promote(zfs_handle_t *zhp)
 		return (zfs_error(hdl, EZFS_BADTYPE, errbuf));
 	}
 
-	(void) strlcpy(parent, zhp->zfs_dmustats.dds_origin, sizeof (parent));
-	if (parent[0] == '\0') {
+	if (zhp->zfs_dmustats.dds_origin[0] == '\0') {
 		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 		    "not a cloned filesystem"));
 		return (zfs_error(hdl, EZFS_BADTYPE, errbuf));
 	}
 
-	(void) strlcpy(zc.zc_value, zhp->zfs_dmustats.dds_origin,
-	    sizeof (zc.zc_value));
-	(void) strlcpy(zc.zc_name, zhp->zfs_name, sizeof (zc.zc_name));
-	ret = zfs_ioctl(hdl, ZFS_IOC_PROMOTE, &zc);
+	ret = lzc_promote(zhp->zfs_name, snapname, sizeof(snapname));
 
 	if (ret != 0) {
 		int save_errno = errno;
```

Comment:
I see that there are some cases where libzfs (incorrectly) uses errno instead of the return value from `lzc_*` (e.g. `zfs_clone()`), but there are other cases where it (correctly) uses the return value (e.g. `zfs_snapshot_nvl()`, `zfs_get_holds()`, `zfs_destroy_snaps_nvl()`).

I don't think that we should add another incorrect use here.  This code should be changed to `switch (ret)`.